lib/vyasa/sangh/sheaf_lattice.ex

```diff
@@ -0,0 +1,159 @@
+defmodule Vyasa.Sangh.SheafLattice do
+  alias Vyasa.Sangh
+  alias Vyasa.Repo
+
+  @moduledoc """
+  A sheaf lattice is a flatmap that represents the tree-structure of sheafs.
+  The keys in this flatmap are the labels for a sheaf and the values are the particular sheafs themselves.
+  In this way, we can keep all of:
+  a) root sheafs: keyed by [x] where is_binary(x)
+  b) level 1 sheafs: keyed by [x, y] where is_binary(x) and is_binary(y)
+  c) level 2 sheafs: keyed by [x, y, z] where is_binary(x) and is_binary(y) and is_binary(z)
+
+  And subsequently, we can read by means of specific filters as well.
+
+  NOTE: this preloads the marks in each sheaf.
+  """
+  def create_complete_sheaf_lattice(sangh_id) when is_binary(sangh_id) do
+    root_sheafs =
+      sangh_id
+      |> Sangh.get_root_sheafs_by_session()
+      |> Enum.filter(fn s -> s.traits == ["published"] end)
+
+    [0, 1, 2]
+    |> Enum.flat_map(fn level ->
+      root_sheafs
+      |> Enum.map(fn sheaf -> to_string(sheaf.path) end)
+      |> Enum.flat_map(fn sheaf_id ->
+        Sangh.get_child_sheafs_by_session(sangh_id, sheaf_id, level)
+      end)
+      |> Enum.map(fn s ->
+        {s.path.labels,
+         s
+         |> Repo.preload(:marks)}
+      end)
+    end)
+    |> Enum.into(%{})
+  end
+
+  @doc """
+  Reads sheaf layers from a lattice based on the specified level and match criteria.
+
+  ## Examples
+
+  # Fetch all sheafs in a particular level
+      iex> SheafLattice.read_sheaf_lattice(sheaf_lattice)
+      # equivalent to:
+      iex> SheafLattice.read_sheaf_lattice(sheaf_lattice, 0, nil)
+
+      iex> SheafLattice.read_sheaf_lattice(sheaf_lattice, 1, nil)
+
+      iex> SheafLattice.read_sheaf_lattice(sheaf_lattice, 2, nil)
+
+  # Fetch based on specific matches of a single label
+      iex> SheafLattice.read_sheaf_lattice(sheaf_lattice, 0, "cf27deab")
+
+      iex> SheafLattice.read_sheaf_lattice(sheaf_lattice, 1, "65c1ac0c")
+
+      iex> SheafLattice.read_sheaf_lattice(sheaf_lattice, 2, "56c369e4")
+
+  # Fetch based on complete matches
+      iex> SheafLattice.read_sheaf_lattice(sheaf_lattice, 1, ["c9cbcb0c", "65c1ac0c"])
+
+      iex> SheafLattice.read_sheaf_lattice(sheaf_lattice, 2, ["c9cbcb0c", "f91bac0d", "56c369e4"])
+
+  # Fetch immediate children based on particular parent
+  # Fetch immediate children of a specific level 0 node:
+      iex> SheafLattice.read_sheaf_lattice(sheaf_lattice, 1, ["cf27deab", nil])
+
+  # Fetch immediate children of a specific level 1 node:
+      iex> SheafLattice.read_sheaf_lattice(sheaf_lattice, 2, ["c9cbcb0c", "65c1ac0c", nil])
+  """
+
+  def read_sheaf_lattice(%{} = sheaf_lattice, level \\ 0, match \\ nil) do
```

Comment:
Here's the main accessor helper function to read from the lattice. My next PR shall be relying heavily on this. 

The examples above show the various cases I expect this read function to be used for.

Reply:
HAHAHA this filter pattern is quite cool create_sheaf_lattice_filter could easily use 
for bot get_in and get_and_update_in behav

[Access.filter(&(create_sheaf_lattice_filter(&1)))]

Reply:
I'll carry on first, will leave it to you to make this change yeah

Reply:
you want me to commit this in(?)